### ISSUE TITLE
Question: What makes DBSP less expressive than Differential Dataflow

### ISSUE THREAD

**Issue by theSherwood** [brurucy/pydbsp#6]
Hi. Thanks for this great repo. I've played around with making a toy implementation of some of Differential Dataflow and then I became aware of the DBSP papers. The approach seems more systematic, but I don't have the background to be able to effectively categorize the differences.

> DBSP is differential dataflow's less expressive successor.

Could you explain why this is the case? I am looking to try to compare and contrast the two approaches.

**Comment by brurucy** [comment]
Hello there @theSherwood.

Thank you so much for checking the library out.

Your issue got buried, sorry to take so long to answer!

Anyhow,
> Could you explain why this is the case?

Sure.

Differential Dataflow (DD) is a programming language, referred to as a "Framework" because in its current form it is embedded in Rust, that extends/is implemented with Timely Dataflow (TD).

TD is a language for computation __distribution__. Think of it as a toolkit to efficiently "distribute" any computation.

DD on the other hand is a high-level language for writing expressive (can recur) incremental programs, where the "incrementality" of the program is made transparent to the user. 

DD's computation graphs are "distributed" with TD. Because of this, DD inherits certain aspects of TD, such as its timestamp/event time sorcery.

DBSP however does not contain any timestamp/event time sorcery. "Time" in DBSP is the time of arrival of some "event", and all "events" are assumed to arrive on time. 

This is a **big** difference with respect to DD, where each "event" may arrive whenever (even in the past).

In short, DBSP is a simplification of DD with most of what has been simplified being TD's inherited semantics.

**Comment by theSherwood** [comment]
@brurucy Thanks for that explanation.

I remember the timestamp sorcery. It seemed like it was making time an N-dimensional space, which was sort of a mismatch for my purposes. And having to manage time in user code was not very ergonomic. I seem to remember the user has to signal when there will be no more events before a certain frontier or something like that so that compaction can take place and certain operators can advance.

Interesting that DBSP does away with all the event time stuff. That seems like it would really simplify implementations. How does it handle iteration/recursion without the timestamps? I seem to remember DD's timestamps being an important part of how that worked, but maybe I've got that wrong.

What I would quite like is an approach in which time is tree-like. It makes time more expensive in practice than what you get with DD's N-dimensional spaces and is suited to a different purpose. Rather than DD's time being useful for coordinating different services, tree-like time would be useful for a git-like/collaborative approach to state. So I'm trying to evaluate whether DD or DBSP is the better starting point.

**Comment by brurucy** [comment]
Hey there @theSherwood, sorry to get to it only now yet again, my GitHub notifications are a mess.

> I remember the timestamp sorcery. It seemed like it was making time an N-dimensional space, which was sort of a mismatch for my purposes. And having to manage time in user code was not very ergonomic. I seem to remember the user has to signal when there will be no more events before a certain frontier or something like that so that compaction can take place and certain operators can advance.

This is indeed a big source of pain with respect to using DDFlow.

You do not have to worry about this in DBSP however, be it this Python implementation of the Rust one. That could also have been solved in the DDFlow library with some clever barrier/semaphoring system (as it is done in the Rust implementation of DBSP iirc).

However, N-dimensional timestamps are very useful and are what makes DDFlow and DBSP fast. If you take a look at how an incremental Join (https://github.com/brurucy/pydbsp/blob/348e814c99b87f5cbe7bcbe00e239bb6cbf7bab2/pydbsp/zset/operators/bilinear.py#L59) is implemented in this library, you will see.

> Interesting that DBSP does away with all the event time stuff. That seems like it would really simplify implementations. How does it handle iteration/recursion without the timestamps? I seem to remember DD's timestamps being an important part of how that worked, but maybe I've got that wrong.

So, it does not do away without the timestamps. It simply restricts them to a total causal order. You can still nest timestamps as much as you want. Take a look at the examples at the very end of this paper: https://arxiv.org/pdf/2203.16684

> What I would quite like is an approach in which time is tree-like. It makes time more expensive in practice than what you get with DD's N-dimensional spaces and is suited to a different purpose. Rather than DD's time being useful for coordinating different services, tree-like time would be useful for a git-like/collaborative approach to state. So I'm trying to evaluate whether DD or DBSP is the better starting point.

I think both could work for this. Give it a go with the library. It is easy (not really) to make time as N-dimensional as you want with `pydbsp`.

**Comment by theSherwood** [comment]
@brurucy Apologies for the late reply. This 

> So, it does not do away without the timestamps. It simply restricts them to a total causal order. You can still nest timestamps as much as you want.

Interesting. I am going to have to dig into this to understand it because I don't yet understand how the timestamps can have total causal order and also support nesting? My understanding is that timestamps with total causal order would most naturally be integers. Or maybe floats if there's some fractional aspect. What is gained by nesting in that case?

**Comment by brurucy** [comment]
> how the timestamps can have total causal order and also support nesting
The timestamps **arrive** in order, but they themselves can be a partial order. By this I meant that in DBSP you cannot have a "earlier" timestamp arrive at a "later" date. This is possible in DBSP however.
